libs/colbert/ragstack_colbert/cassandra_database.py

```
@@ -115,55 +130,158 @@ def add_chunks(self, chunks: List[Chunk]) -> List[Tuple[str, int]]:
+                text=text,
                 metadata=metadata,
+                vector=vector,
+            )
+            future.add_callbacks(
+                callback=_result_callback,
+                callback_args=(task_id,),
+                errback=_error_callback,
+                errback_args=(task_id,),
             )
 
-            futures.append((doc_id, chunk_id, -1, future))
+        for chunk in chunks:
+            doc_id = chunk.doc_id
+            chunk_id = chunk.chunk_id
+            text = chunk.text
+            metadata = chunk.metadata
+
+            _init_put(
+                doc_id=doc_id,
+                chunk_id=chunk_id,
+                embedding_id=-1,
+                text=text,
+                metadata=metadata,
+            )
 
             for index, vector in enumerate(chunk.embedding):
-                future = self._table.put_async(
-                    partition_id=doc_id, row_id=(chunk_id, index), vector=vector
+                _init_put(
+                    doc_id=doc_id, chunk_id=chunk_id, embedding_id=index, vector=vector
                 )
-                futures.append((doc_id, chunk_id, index, future))
-
-            results: List[Tuple[str, int]] = []
-            for doc_id, chunk_id, embedding_id, future in futures:
-                try:
-                    future.result()
-                    results.append((doc_id, chunk_id))
-                except Exception as e:
-                    if embedding_id == -1:
-                        logging.error(
-                            f"issue inserting document data: {doc_id} chunk: {chunk_id}: {e}"
-                        )
-                    else:
-                        logging.error(
-                            f"issue inserting document embedding: {doc_id} chunk: {chunk_id} embedding: {embedding_id}: {e}"
-                        )
+
+        delay = 0.05  # Start with a 50 ms delay
+        while len(all_tasks) > 0:
```

shall we raise an error if one or more tasks failed ?

---

I'll add that